File: java/src/main/java/com/penapereira/patterns/objectpool/ObjectPool.java
package com.penapereira.patterns.objectpool;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.LinkedHashSet;
import java.util.Set;

/**
 * Lends out up to {@code capacity} reusable objects, creating each one only on
 * first demand and recycling released ones in FIFO order.
 */
public final class ObjectPool {

    private final int capacity;
    private final Deque<Reusable> available = new ArrayDeque<>();
    private final Set<Reusable> inUse = new LinkedHashSet<>();
    private int created;

    public ObjectPool(int capacity) {
        if (capacity < 1) {
            throw new IllegalArgumentException("capacity must be positive: " + capacity);
        }
        this.capacity = capacity;
    }

    /**
     * Hands out an idle object, or a new one while the pool is below capacity.
     *
     * @throws IllegalStateException when every object is in use
     */
    public Reusable acquire(String task) {
        Reusable reusable = available.pollFirst();
        if (reusable == null) {
            if (created == capacity) {
                throw new IllegalStateException("pool exhausted, " + inUse.size() + " of " + capacity + " in use");
            }
            created++;
            reusable = new Reusable(created);
        }
        reusable.assign(task);
        inUse.add(reusable);
        return reusable;
    }

    /**
     * Takes an object back, resets it and makes it available again.
     *
     * @throws IllegalArgumentException when the object was not lent out by this pool
     */
    public void release(Reusable reusable) {
        if (!inUse.remove(reusable)) {
            throw new IllegalArgumentException(reusable + " is not in use");
        }
        reusable.reset();
        available.addLast(reusable);
    }

    public int capacity() {
        return capacity;
    }

    public int created() {
        return created;
    }

    public int available() {
        return available.size();
    }

    public int inUse() {
        return inUse.size();
    }
}

File: javascript/src/object-pool/object-pool.js
import { Reusable } from "./reusable.js";

/**
 * Lends out up to `capacity` reusable objects, creating each one only on
 * first demand and recycling released ones in FIFO order.
 */
export class ObjectPool {
  #capacity;
  #idle = [];
  #lent = new Set();
  #created = 0;

  constructor(capacity) {
    if (capacity < 1) {
      throw new Error(`capacity must be positive: ${capacity}`);
    }
    this.#capacity = capacity;
  }

  /**
   * Hands out an idle object, or a new one while the pool is below capacity.
   * @throws Error when every object is in use
   */
  acquire(task) {
    let reusable = this.#idle.shift();
    if (reusable === undefined) {
      if (this.#created === this.#capacity) {
        throw new Error(`pool exhausted, ${this.#lent.size} of ${this.#capacity} in use`);
      }
      this.#created += 1;
      reusable = new Reusable(this.#created);
    }
    reusable.assign(task);
    this.#lent.add(reusable);
    return reusable;
  }

  /**
   * Takes an object back, resets it and makes it available again.
   * @throws Error when the object was not lent out by this pool
   */
  release(reusable) {
    if (!this.#lent.delete(reusable)) {
      throw new Error(`${reusable} is not in use`);
    }
    reusable.reset();
    this.#idle.push(reusable);
  }

  get capacity() {
    return this.#capacity;
  }

  get created() {
    return this.#created;
  }

  get available() {
    return this.#idle.length;
  }

  get inUse() {
    return this.#lent.size;
  }
}
